hardware_interface/include/hardware_interface/robot_hw.h

```diff
@@ -51,8 +51,18 @@ namespace hardware_interface
  * hardware.
  *
  * The hardware interface map (\ref interfaces_) is a 1-to-1 map between
- * the names of interface types derived from \ref HardwareInterface  and
+ * the names of interface types derived from \ref HardwareInterface and
  * instances of those interface types.
+ * 
+ * The registration (\ref registerInterface) of interfaces can be done in the 
+ * constructor or init() of a custom robot hardware class.
+ * 
+ * The robot state is described by data members of a class derived from this 
+ * base interface. The names of these data members should give semantic meaning 
+ * (pos, vel, eff) to the registered resources (joints, sensors, actuators). 
+ * Commands from controllers are represented by a member array (e.g. cmd)
+ * and populated by the controller's update method, 
+ * (controller_interface::ControllerBase::update()).
```

Comment:
Not sure if this gives too much details on how to implement a derived class. If this is the wrong place, I think there should be a link to an example showing how it should be done. For example to the [wiki page](https://github.com/ros-controls/ros_control/wiki/hardware_interface#setting-up-a-new-robot). But as I've mentioned, I think this example should be updated.

Reply:
IMO this is still too much detail about implementation, I don't really like recommending data types and variable names and that sort of thing, but mentioning that this class stores the state of the robot's hardware resources and outgoing commands is fine. Maybe it should live up in the first paragraph, to elaborate on "maintains a map of the hardware interfaces" (Line 49)?

Reply:
I agree, this is probably not the best place for data type and variable name suggestions and I'll move this part up to the class description (Line 49). 

What about updating the [example in the wiki page](https://github.com/ros-controls/ros_control/wiki/hardware_interface#setting-up-a-new-robot)? I think this could be another option: we extend the example slightly (if required) and link to it in the class description (somewhere around line 49). Another way I could imagine is to put the wiki example into this class description. This is also done in the detailed description of the [transmission_interface](http://docs.ros.org/melodic/api/transmission_interface/html/c++/classtransmission__interface_1_1TransmissionInterfaceLoader.html#details).

Reply:
I like the idea of putting a minimal example (such as the wiki example you linked) in the class description, that's a lot less fragile than linking to the wiki since the wiki could always change and break the link.

Reply:
As mentioned below, I've put the wiki example (slightly modified) into a new `README.md` which acts as doxygen's mainpage and also as readme for the `hardware_interface` folder of the github repository.